src/main/java/org/broadinstitute/hellbender/utils/locusiterator/AlignmentContextIteratorBuilder.java

```diff
@@ -0,0 +1,178 @@
+     * @param isIncludeNs include reads with N on the loci in question
+     * @return iterator that produces AlignmentContexts ready for consumption (e.g. by a {@link org.broadinstitute.hellbender.engine.LocusWalker})
+     */
+    private static Iterator<AlignmentContext> createAlignmentContextIterator(final List<SimpleInterval> intervalsForTraversal,
+                                                                            final SAMFileHeader header,
+                                                                               final Iterator<GATKRead> readIterator,
+                                                                               final SAMSequenceDictionary dictionary,
+                                                                               final LIBSDownsamplingInfo downsamplingInfo,
+                                                                               final boolean isReference,
+                                                                               boolean emitEmptyLoci,
+                                                                               boolean isKeepUniqueReadListInLibs,
+                                                                               boolean isIncludeDeletions,
+                                                                               boolean isIncludeNs) {
+
+        // get the samples from the read groups
+        final Set<String> samples = header.getReadGroups().stream()
+                .map(SAMReadGroupRecord::getSample)
+                .collect(Collectors.toSet());
+
+        // get the LIBS
+        final LocusIteratorByState libs = new LocusIteratorByState(readIterator, downsamplingInfo, isKeepUniqueReadListInLibs, samples, header, isIncludeDeletions, isIncludeNs);
+        Iterator<AlignmentContext> iterator;
+
+        List<SimpleInterval> finalIntervals = intervalsForTraversal;
+        validateEmitEmptyLociParameters(emitEmptyLoci, dictionary, intervalsForTraversal, isReference);
+        if (emitEmptyLoci) {
+
+            // If no intervals were specified, then use the entire reference (or best available sequence dictionary).
+            if (!hasIntervals(finalIntervals)) {
+                finalIntervals = IntervalUtils.getAllIntervalsForReference(dictionary);
+            }
+            final IntervalLocusIterator intervalLocusIterator = new IntervalLocusIterator(finalIntervals.iterator());
+            iterator =  new IntervalAlignmentContextIterator(libs, intervalLocusIterator, header.getSequenceDictionary());
+
+        } else {
+            // prepare the iterator
+            iterator = hasIntervals(finalIntervals) ? new IntervalOverlappingIterator<>(libs, finalIntervals, header.getSequenceDictionary()) : libs;
+        }
+
+        return iterator;
+    }
+
+    private static boolean hasIntervals(final List<SimpleInterval> finalIntervals) {
+        return finalIntervals != null;
```

Comment:
This means that if the intervals are empty, the iterator has intervals...is that what we want or should it account for an empty list too?